src/librustc_metadata/encoder.rs

```diff
@@ -833,6 +833,12 @@ impl<'a, 'b: 'a, 'tcx: 'b> IsolatedEncoder<'a, 'b, 'tcx> {
         }
     }
 
+    fn metadata_output_only(&self) -> bool {
+        // MIR optimisation can be skipped when we're just interested in the metadata.
+        self.tcx.sess.opts.output_types.len() == 1 &&
```

Comment:
Cargo also includes dep-info as output; we probably need to check if we have anything any output that requires codegen.